File: haskell/src/Chapter3.hs
module Chapter3 where
import Prelude hiding (tail,
                       dropWhile,
                       takeWhile,
                       drop,
                       init,
                       length,
                       reverse,
                       foldl,
                       concat,
                       map,
                       filter, zipWith)


data List a = Empty | Cons a (List a) deriving (Eq, Show)

-- Exercise 3.2
tail :: (List a) -> (List a)
tail Empty = Empty
tail (Cons _ t) = t

-- Exercise 3.3
setHead :: a -> (List a) -> List a
setHead s Empty = Cons s Empty
setHead s (Cons h t) = Cons s t

-- Exercise 3.4
drop :: Int -> (List a) -> (List a)
drop _ Empty = Empty
drop n (Cons h t)
     | n <= 0 = (Cons h t)
     | otherwise = drop (n - 1) t

-- tail as drop
tail2 :: (List a) -> (List a)
tail2  = drop 1

-- Exercise 3.5
dropWhile :: (a -> Bool) -> (List a) -> (List a)
dropWhile _ Empty = Empty
dropWhile f (Cons h t) = if( f h ) then dropWhile f t else (Cons h t)


-- Exercise 3.6
init ::  (List a) -> (List a)
init Empty = Empty
init (Cons x Empty) = Empty
init (Cons h t) = (Cons h (init t))


takeWhile :: (a -> Bool) -> (List a) -> (List a)
takeWhile _ Empty = Empty
takeWhile f (Cons h t) = if (f h) then (Cons h (takeWhile f t)) else Empty

-- Exercise 3.7
foldRight :: ( a -> b -> b ) -> b -> (List a) -> b
foldRight _ b Empty = b
foldRight f b (Cons h t) = f h (foldRight f b t)

-- Exercise 3.8 is already covered by 3.7

-- Exercise 3.9
length :: (List a) -> Int
length xs = foldRight (\x t -> t + 1) 0 xs


-- Exercise 3.10
foldLeft :: (b -> a -> b) -> b -> (List a) -> b
foldLeft _ b Empty = b
foldLeft f b (Cons h t) = let b' = b `f` h
                     in seq b' $ foldLeft f b' t

-- Exercise 3.11
sum :: (Num a) => (List a) -> a
sum = foldLeft (\t x -> x + t) 0

-- Exercise 3.11
product :: (Num a) => (List a) -> a
product = foldLeft (\t x -> x * t) 1


-- Exercise 3.11
length2 :: (Num b) => (List a) -> b
length2 = foldLeft (\t x -> t + 1) 0


-- Exercise 3.12
reverse :: (List a) -> (List a)
reverse = foldLeft (\t x -> Cons x t ) (Empty)


-- Exercise 3.13
foldLeftWithFoldRight :: (b -> a -> b) -> b -> (List a) -> b
foldLeftWithFoldRight f b l = foldRight (\x g t -> g ( f t x ) ) id l b

-- Exercise 3.13
foldRightWithFoldLeft :: (a -> b -> b) -> b -> (List a) -> b
foldRightWithFoldLeft f b l = foldLeft (\g x t ->  g(f x t) ) id l b

-- Exercise 3.14
append :: (List a) -> (List a) -> (List a)
append Empty b = b
append a Empty = a
append a b = foldRight (\x t -> Cons x t ) b a


-- Exercise 3.15
concat :: (List(List a)) -> (List a)
concat Empty = Empty
concat xs  = foldRight (\x t -> append x t) Empty xs

-- Exercise 3.16, 3.17 and 3.18
map :: (a -> b) ->(List a) -> (List b)
map f Empty = Empty
map f (Cons h t) = Cons (f h) (map f t)

-- Exercise 3.19
filter :: (a -> Bool) -> (List a) -> (List a)
filter p ls = foldRight (\h t -> if ( p h ) then Cons h t else t  ) Empty ls

-- Exercise 3.20
flatMap :: (a -> List b) -> List a -> List b
flatMap f l = concat (map f l)


-- Exercise 3.21
filterUsingFlatMap  :: (a -> Bool) -> (List a) -> (List a)
filterUsingFlatMap f l = flatMap (\a -> if (f a) then Cons a Empty else Empty) l

-- Exercise 3.22 and 3.23
zipWith :: (a -> b -> c) -> (List a) -> (List b) -> (List c)
zipWith _ _ Empty = Empty
zipWith _ Empty _ = Empty
zipWith f (Cons h1 t1) (Cons h2 t2) = Cons (f h1 h2) (zipWith f t1 t2)

File: scala/src/main/scala/com/ad/Chapter3.scala
package com.ad

object Chapter3 {
  sealed trait List[+A]
  case class Cons[+A](head: A, tail: List[A]) extends List[A]
  case object Nil extends List[Nothing]


  //3.2
  def tail[A](l: List[A] ): List[A] = l match {
    case Nil => throw new IllegalArgumentException("Cannot call tail on an empty list.")
    case Cons(h , t) => t
  }

  //3.3
  def setHead[A](a: A , l: List[A]) : List[A] = l match {
    case Nil => Cons(a , Nil)
    case Cons(h, t) => Cons(a, t)
  }
  //3.4
  def drop[A](l: List[A], n:Int ) : List[A] = (n, l) match {
    case (_, Nil) => Nil
    case (0,  Cons(h , t)) => Cons(h , t)
    case (i,  Cons(h , t)) => drop(t, i - 1)
  }

  //3.4
  def tailWithDrop[A](l: List[A]) : List[A] = l match {
    case Nil => throw new IllegalArgumentException ("tail on an empty list")
    case xs => drop(xs , 1)
  }

  //3.5
  def dropWhile[A](l: List[A], f:A => Boolean ) : List[A] = l match {
    case Nil => Nil
    case Cons(h, t) if f(h) => dropWhile(t , f)
    case ls => ls
  }


  // 3.6
  def init[A](l: List[A]) : List[A] = l match {
    case Nil => Nil
    case Cons(h, Nil) => Nil
    case Cons(h, t) => Cons(h , init(t))
  }


  def takeWhile[A](l: List[A]) (f: A => Boolean) : List[A] = l match {
    case Nil => Nil
    case Cons(h, t) => if( f(h) ) Cons( h , takeWhile(t)(f)) else Nil
  }

  //  -- Exercise 3.7
  def foldRight[A, B](l: List[A])(z: B)(f:(A, B) => B): B = (l, z) match {
    case (Nil, b) => b
    case (Cons(h, t), b) => f (h, foldRight(t)(b)(f) )
  }

  // Exercise 3.8 is already covered by 3.7

  // Exercise 3.9
  def lengthWithFoldRight[A](l : List[A]) : Int = {
    foldRight(l)(0){(x, t:Int) => 1 + t}
  }

  // Exercise 3.10
  def foldLeft[A,B](l: List[A])(z: B)(f: (B, A) => B) : B = l match {
    case Nil => z
    case Cons (h, t) =>  foldLeft(t)(f(z,h))(f)
  }

  //  -- Exercise 3.11
  def sum(l: List[Int]) : Int = {
    foldLeft(l)(0){(a,b) => a + b}
  }

  def product(l: List[Int]) : Int = {
    foldLeft(l)(1){(a,b) => a * b}
  }

  def length[A](l: List[A]) : Int = {
    foldLeft(l)(0){(a,b) => a + 1}
  }

  def reverse[A](l: List[A]) : List[A]= {
    foldLeft(l)(Nil:List[A]){(a,b) => Cons(b, a)}
  }


  def foldLeftWithFoldRight[A,B](l: List[A])(z: B)(f: (B, A) => B) : B = {
    foldRight(l){i:B => i}{(x:A, g) => {(t:B) => g(f(t , x))} }(z)
  }

  def append[A](to: List[A] , from:List[A]) : List[A] = {
    foldRight(to)(from){(x,t) => Cons(x, t)}
  }

  def concat[A](l: List[List[A]] ) : List[A] = {
    foldRight(l)(Nil:List[A]){(x,t) => append(x,t)}
  }

  // Exercise 3.16, 3.17 and 3.18
  def map[A,B](l : List[A])(f:A => B):List[B] = {
    foldRight(l)(Nil:List[B]){(x,t) => Cons(f(x) , t)}
  }


  // Exercise 3.19
  def filter[A](as: List[A])(f: A => Boolean): List[A] = {
    foldRight(as)(Nil:List[A]){(x,t) => if( f(x) )Cons(x, t) else t}
  }


  // Exercise 3.20
  def flatMap[A](as: List[A])(f: A => List[A]): List[A] = {
    foldRight(as)(Nil:List[A]){(x,t) => append(f(x),t)}
  }

  // Exercise 3.21
  def filterwithFlatMap[A](as: List[A])(f: A => Boolean): List[A] = {
    flatMap(as)((x) => if(f(x)) Cons(x,Nil) else Nil )
  }


  // Exercise 3.22, 3.23
  def zipWith[A,B,C](as:List[A], bs:List[B])(combine:(A,B)=>C) : List[C] = (as, bs) match {
    case (Nil, _) => Nil
    case (_, Nil) => Nil
    case (Cons(x,xs) , Cons(y, ys)) => Cons(combine(x,y) , zipWith(xs,ys)(combine) )
  }

  //Exercise 3.24
  def hasSubsequence[A](sup: scala.List[A], sub: scala.List[A]): Boolean = {
    sup.inits.flatMap(_.tails).filter(_.nonEmpty). contains(sub)
  }

  // Exercise 3.25-3.28

  sealed trait Tree[+A]
  case class Leaf[A](value: A) extends Tree[A]
  case class Branch[A](left: Tree[A], right: Tree[A]) extends Tree[A]

  def size[A](t:Tree[A]) : Int = t match {
    case Leaf(_) => 1
    case (Branch(l,r))  => 1 + size(l) + size(r)
  }

  def maximum(t:Tree[Int]) : Int = t match {
    case Leaf(v) => v
    case (Branch(l,r))  => maximum(l) max maximum(r)
  }

  def depth(t:Tree[Int]) : Int = t match {
    case Leaf(v) => 1
    case (Branch(l,r))  =>depth(l) max depth(r)
  }


  def fold[A,B](t:Tree[A])(f:A => B )(g:(B,B) => B) : B = t match {
    case Leaf(v) => f(v)
    case (Branch(l,r)) => g(fold(l)(f)(g) , fold(r)(f)(g))
  }

  def sizeWithFold[A](t:Tree[A]): Int = {
    fold(t)(_ => 1)(1 + _ + _)
  }

  def depthWithFold[A](t:Tree[A]): Int = {
    fold(t)(_ => 1 )( (l,r) => (l + 1) max (r + 1))
  }

  def mapWithFold[A,B](t:Tree[A])(f: A => B)(g:(B , B) => B) : B = {
    fold(t)(f)(g)
  }



}